File: backend/app/schemas/qc.py
from datetime import datetime

from pydantic import BaseModel, field_validator


# ---------------------------------------------------------------------------
# Defect Type schemas
# ---------------------------------------------------------------------------


class DefectTypeCreate(BaseModel):
    defect_name: str
    product_category: str
    default_severity: str = "minor"
    default_disposition: str = "hold_pending_review"
    description: str | None = None


class DefectTypeResponse(BaseModel):
    id: str
    company_id: str
    defect_name: str
    product_category: str
    default_severity: str
    default_disposition: str
    description: str | None = None
    is_active: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Inspection Step schemas
# ---------------------------------------------------------------------------


class StepCreate(BaseModel):
    step_name: str
    step_order: int = 0
    inspection_type: str = "visual"
    description: str | None = None
    pass_criteria: str | None = None
    photo_required: bool = False
    required: bool = True


class StepUpdate(BaseModel):
    step_name: str | None = None
    step_order: int | None = None
    inspection_type: str | None = None
    description: str | None = None
    pass_criteria: str | None = None
    photo_required: bool | None = None
    required: bool | None = None

    @field_validator("*", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: object) -> object:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class StepResponse(BaseModel):
    id: str
    template_id: str
    company_id: str
    step_name: str
    step_order: int
    inspection_type: str
    description: str | None = None
    pass_criteria: str | None = None
    photo_required: bool
    required: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Template schemas
# ---------------------------------------------------------------------------


class TemplateCreate(BaseModel):
    product_category: str
    template_name: str
    description: str | None = None
    wilbert_warranty_compliant: bool = False
    steps: list[StepCreate] | None = None


class TemplateResponse(BaseModel):
    id: str
    company_id: str
    product_category: str
    template_name: str
    description: str | None = None
    wilbert_warranty_compliant: bool
    is_active: bool
    steps: list[StepResponse] = []
    created_at: datetime
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Step Result schemas
# ---------------------------------------------------------------------------


class StepResultUpdate(BaseModel):
    result: str
    notes: str | None = None
    defect_type_id: str | None = None
    defect_severity: str | None = None

    @field_validator("*", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: object) -> object:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class StepResultResponse(BaseModel):
    id: str
    inspection_id: str
    step_id: str
    company_id: str
    result: str
    notes: str | None = None
    defect_type_id: str | None = None
    defect_severity: str | None = None
    created_at: datetime
    step: StepResponse | None = None
    defect_type: DefectTypeResponse | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Media schemas
# ---------------------------------------------------------------------------


class MediaCreate(BaseModel):
    inspection_id: str
    step_result_id: str | None = None
    file_url: str
    caption: str | None = None
    captured_at: datetime | None = None


class MediaResponse(BaseModel):
    id: str
    step_result_id: str | None = None
    inspection_id: str
    company_id: str
    file_url: str
    caption: str | None = None
    captured_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Disposition schemas
# ---------------------------------------------------------------------------


class DispositionCreate(BaseModel):
    disposition: str
    disposition_notes: str | None = None
    rework_instructions: str | None = None


class DispositionResponse(BaseModel):
    id: str
    inspection_id: str
    company_id: str
    decided_by: str
    decided_by_name: str | None = None
    disposition: str
    disposition_notes: str | None = None
    rework_instructions: str | None = None
    decided_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Rework schemas
# ---------------------------------------------------------------------------


class ReworkCreate(BaseModel):
    rework_description: str


class ReworkResponse(BaseModel):
    id: str
    inspection_id: str
    original_inspection_id: str
    company_id: str
    rework_description: str
    rework_completed_by: str | None = None
    rework_completed_at: datetime | None = None
    re_inspection_id: str | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Inspection schemas
# ---------------------------------------------------------------------------


class InspectionCreate(BaseModel):
    inventory_item_id: str | None = None
    template_id: str
    product_type: str | None = None


class InspectionResponse(BaseModel):
    id: str
    company_id: str
    inventory_item_id: str | None = None
    template_id: str
    product_category: str
    product_type: str | None = None
    inspector_id: str
    inspector_name: str | None = None
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    overall_notes: str | None = None
    certificate_number: str | None = None
    created_at: datetime
    updated_at: datetime | None = None
    step_results: list[StepResultResponse] = []
    disposition: DispositionResponse | None = None
    media: list[MediaResponse] = []

    model_config = {"from_attributes": True}


class InspectionListItem(BaseModel):
    id: str
    company_id: str
    inventory_item_id: str | None = None
    template_id: str
    product_category: str
    product_type: str | None = None
    inspector_id: str
    inspector_name: str | None = None
    template_name: str | None = None
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    certificate_number: str | None = None
    created_at: datetime
    pass_count: int = 0
    fail_count: int = 0
    step_count: int = 0

    model_config = {"from_attributes": True}


class InspectionListResponse(BaseModel):
    items: list[InspectionListItem]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# Report schemas
# ---------------------------------------------------------------------------


class DefectFrequency(BaseModel):
    defect_name: str
    product_category: str
    count: int
    severity: str


class InspectorPerformance(BaseModel):
    inspector_id: str
    inspector_name: str
    total_inspections: int
    pass_count: int
    fail_count: int
    avg_duration_minutes: float | None = None


class QCSummaryReport(BaseModel):
    total_inspections: int
    pass_rate: float
    fail_rate: float
    conditional_pass_rate: float
    rework_rate: float
    rework_success_rate: float
    avg_time_in_qc_minutes: float | None = None
    defect_frequency: list[DefectFrequency]
    inspector_performance: list[InspectorPerformance]


# ---------------------------------------------------------------------------
# Dashboard schemas
# ---------------------------------------------------------------------------


class DashboardStats(BaseModel):
    pending_count: int
    in_progress_count: int
    failed_today_count: int
    passed_today_count: int
    rework_pending_count: int


# ---------------------------------------------------------------------------
# Complete inspection request
# ---------------------------------------------------------------------------


class CompleteInspectionRequest(BaseModel):
    overall_notes: str | None = None


# ---------------------------------------------------------------------------
# Paginated inspections
# ---------------------------------------------------------------------------


class PaginatedInspections(BaseModel):
    items: list[InspectionListItem]
    total: int
    page: int
    per_page: int

File: frontend/src/types/qc.ts
export type ProductCategory = "burial_vault" | "columbarium" | "monument" | "redi_rock" | "precast_other";
export type InspectionStatus = "pending" | "in_progress" | "passed" | "failed" | "conditional_pass" | "rework_required";
export type StepResult = "pending" | "pass" | "fail" | "na";
export type DefectSeverity = "minor" | "major" | "critical";
export type DispositionType = "scrap" | "rework" | "conditional_pass" | "hold_pending_review";
export type InspectionType = "visual" | "pressure_test" | "dimensional" | "photo_required";

export interface QCTemplate {
  id: string;
  company_id: string;
  name: string;
  product_category: ProductCategory;
  description: string | null;
  is_active: boolean;
  steps: QCStep[];
  created_at: string;
  updated_at: string;
}

export interface QCStep {
  id: string;
  template_id: string;
  step_number: number;
  name: string;
  description: string | null;
  inspection_type: InspectionType;
  is_required: boolean;
  photo_required: boolean;
  sort_order: number;
}

export interface QCInspection {
  id: string;
  company_id: string;
  inspection_number: string;
  template_id: string;
  template_name: string;
  inventory_item_id: string | null;
  item_identifier: string;
  product_category: ProductCategory;
  status: InspectionStatus;
  inspector_id: string;
  inspector_name: string;
  started_at: string | null;
  completed_at: string | null;
  overall_notes: string | null;
  step_results: QCStepResult[];
  dispositions: QCDisposition[];
  rework_records: QCReworkRecord[];
  created_at: string;
  updated_at: string;
}

export interface QCStepResult {
  id: string;
  inspection_id: string;
  step_id: string;
  step_number: number;
  step_name: string;
  step_description: string | null;
  inspection_type: InspectionType;
  photo_required: boolean;
  result: StepResult;
  defect_type_id: string | null;
  defect_type_name: string | null;
  defect_severity: DefectSeverity | null;
  notes: string | null;
  media: QCMedia[];
}

export interface QCDefectType {
  id: string;
  company_id: string;
  name: string;
  description: string | null;
  product_category: ProductCategory | null;
  default_severity: DefectSeverity;
  is_active: boolean;
}

export interface QCDisposition {
  id: string;
  inspection_id: string;
  disposition_type: DispositionType;
  reason: string | null;
  rework_instructions: string | null;
  decided_by_id: string;
  decided_by_name: string;
  created_at: string;
}

export interface QCReworkRecord {
  id: string;
  inspection_id: string;
  disposition_id: string;
  assigned_to_id: string | null;
  assigned_to_name: string | null;
  instructions: string | null;
  status: "pending" | "in_progress" | "completed" | "failed";
  started_at: string | null;
  completed_at: string | null;
  result_notes: string | null;
  created_at: string;
}

export interface QCMedia {
  id: string;
  step_result_id: string;
  file_url: string;
  file_name: string;
  content_type: string;
  created_at: string;
}

// Create / Update types

export interface InspectionCreate {
  item_identifier: string;
  product_category: ProductCategory;
  template_id?: string;
  inventory_item_id?: string;
}

export interface StepResultUpdate {
  result: StepResult;
  defect_type_id?: string;
  defect_severity?: DefectSeverity;
  notes?: string;
}

export interface DispositionCreate {
  disposition_type: DispositionType;
  reason?: string;
  rework_instructions?: string;
}

export interface ReworkCreate {
  disposition_id: string;
  assigned_to_id?: string;
  instructions?: string;
}

// Report types

export interface QCSummaryReport {
  pass_rate_by_category: Record<string, { total: number; passed: number; rate: number }>;
  defect_frequency: { defect_type: string; count: number; severity: DefectSeverity }[];
  inspector_stats: { inspector_name: string; total: number; pass_rate: number }[];
  rework_success_rate: number;
  avg_time_in_qc: Record<string, number>;
}

// List types

export interface QCInspectionListItem {
  id: string;
  inspection_number: string;
  item_identifier: string;
  product_category: ProductCategory;
  status: InspectionStatus;
  inspector_name: string;
  started_at: string | null;
  completed_at: string | null;
  step_count: number;
  pass_count: number;
  fail_count: number;
  created_at: string;
}

export interface PaginatedInspections {
  items: QCInspectionListItem[];
  total: number;
  page: number;
  per_page: number;
}

export interface QCDashboardStats {
  pending_count: number;
  in_rework_count: number;
  awaiting_disposition_count: number;
  completed_today_count: number;
}
